File: Tests/GlasstualTests/IRCModelMigrationTests.m
/* *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */

#import <XCTest/XCTest.h>

#import "GLTTestClient.h"
#import "IRCChannelConfig.h"
#import "IRCConnection.h"
#import "IRCConnectionConfig.h"
#import "IRCConnectionPrivate.h"
#import "IRCHighlightLogEntryPrivate.h"
#import "IRCHighlightMatchCondition.h"
#import "IRCServerPrivate.h"
#import "IRCUserPersistentStorePrivate.h"
#import "IRCUserRelationsPrivate.h"
#import "TVCLogLinePrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface IRCModelMigrationTests : XCTestCase
@end

@implementation IRCModelMigrationTests

#pragma mark - Connection

- (void)testConnectionInitialStateAndConfigIsolation
{
	GLTTestClient *client = [GLTTestClient testClient];
	IRCConnectionConfigMutable *sourceConfig = [IRCConnectionConfigMutable new];
	sourceConfig.serverAddress = @"irc.example.test";

	IRCConnection *connection = [[IRCConnection alloc] initWithConfig:sourceConfig onClient:client];
	sourceConfig.serverAddress = @"changed.example.test";

	XCTAssertEqual(connection.client, client);
	XCTAssertEqualObjects(connection.config.serverAddress, @"irc.example.test");
	XCTAssertGreaterThan(connection.uniqueIdentifier.length, 0);
	XCTAssertFalse(connection.isConnecting);
	XCTAssertFalse(connection.isConnected);
	XCTAssertFalse(connection.isDisconnecting);
	XCTAssertFalse(connection.isSecured);
	XCTAssertFalse(connection.isSending);
	XCTAssertFalse(connection.EOFReceived);
}

- (void)testConnectionResetClearsTransientState
{
	GLTTestClient *client = [GLTTestClient testClient];
	IRCConnection *connection = [[IRCConnection alloc] initWithConfig:[IRCConnectionConfig new] onClient:client];

	[connection resetState];

	XCTAssertFalse(connection.isConnecting);
	XCTAssertFalse(connection.isConnected);
	XCTAssertFalse(connection.isConnectedWithClientSideCertificate);
	XCTAssertFalse(connection.isDisconnecting);
	XCTAssertFalse(connection.isSecured);
	XCTAssertFalse(connection.isSending);
	XCTAssertFalse(connection.EOFReceived);
	XCTAssertNil(connection.connectedAddress);
}

#pragma mark - Channel config

- (void)testChannelConfigDefaultsAndLegacyKeys
{
	IRCChannelConfig *defaults = [IRCChannelConfig seedWithName:@"#general"];

	XCTAssertTrue(defaults.autoJoin);
	XCTAssertTrue(defaults.pushNotifications);
	XCTAssertTrue(defaults.showTreeBadgeCount);
	XCTAssertEqual(defaults.type, IRCChannelTypeChannel);
	XCTAssertEqualObjects(defaults.channelName, @"#general");
	XCTAssertGreaterThan(defaults.uniqueIdentifier.length, 0);

	IRCChannelConfig *legacy = [[IRCChannelConfig alloc] initWithDictionary:@{
		@"channelName" : @"#legacy",
		@"joinOnConnect" : @NO,
		@"ignoreJPQActivity" : @YES,
		@"enableNotifications" : @NO,
		@"enableTreeBadgeCountDrawing" : @NO
	}];

	XCTAssertFalse(legacy.autoJoin);
	XCTAssertTrue(legacy.ignoreGeneralEventMessages);
	XCTAssertFalse(legacy.pushNotifications);
	XCTAssertFalse(legacy.showTreeBadgeCount);
}

- (void)testChannelConfigMutableAndUniqueCopiesPreserveValues
{
	IRCChannelConfigMutable *mutable = [IRCChannelConfigMutable new];
	mutable.channelName = @"#swift";
	mutable.label = @"Swift migration";
	mutable.defaultModes = @"+nt";
	mutable.secretKey = @"join-key";

	IRCChannelConfig *copy = [mutable copy];
	IRCChannelConfigMutable *unique = [mutable uniqueCopyMutable];

	XCTAssertEqualObjects(copy.channelName, @"#swift");
	XCTAssertEqualObjects(copy.label, @"Swift migration");
	XCTAssertEqualObjects(copy.defaultModes, @"+nt");
	XCTAssertEqualObjects(copy.secretKey, @"join-key");
	XCTAssertEqualObjects(copy.uniqueIdentifier, mutable.uniqueIdentifier);

	XCTAssertEqualObjects(unique.channelName, @"#swift");
	XCTAssertEqualObjects(unique.secretKey, @"join-key");
	XCTAssertNotEqualObjects(unique.uniqueIdentifier, mutable.uniqueIdentifier);
}

- (void)testChannelConfigNotificationOverridesUseThreeStateSemantics
{
	IRCChannelConfigMutable *config = [IRCChannelConfigMutable new];
	TXNotificationType event = TXNotificationTypeHighlight;

	XCTAssertEqual([config notificationEnabledForEvent:event], NSControlStateValueMixed);

	[config setNotificationEnabled:NSControlStateValueOn forEvent:event];
	XCTAssertEqual([config notificationEnabledForEvent:event], NSControlStateValueOn);

	[config setNotificationEnabled:NSControlStateValueOff forEvent:event];
	XCTAssertEqual([config notificationEnabledForEvent:event], NSControlStateValueOff);

	[config setNotificationEnabled:NSControlStateValueMixed forEvent:event];
	XCTAssertEqual([config notificationEnabledForEvent:event], NSControlStateValueMixed);
}

#pragma mark - Highlight match condition

- (void)testHighlightMatchConditionRoundTripsDictionaryAndDefaults
{
	IRCHighlightMatchCondition *condition = [[IRCHighlightMatchCondition alloc]
		initWithDictionary:@{@"matchKeyword" : @"alert", @"matchChannelID" : @"chan-1", @"matchIsExcluded" : @YES}];

	XCTAssertEqualObjects(condition.matchKeyword, @"alert");
	XCTAssertEqualObjects(condition.matchChannelId, @"chan-1");
	XCTAssertTrue(condition.matchIsExcluded);
	XCTAssertGreaterThan(condition.uniqueIdentifier.length, 0);

	NSDictionary *dictionary = condition.dictionaryValue;
	XCTAssertEqualObjects(dictionary[@"matchKeyword"], @"alert");
	XCTAssertEqualObjects(dictionary[@"matchChannelID"], @"chan-1");
	XCTAssertEqualObjects(dictionary[@"matchIsExcluded"], @YES);
	XCTAssertEqualObjects(dictionary[@"uniqueIdentifier"], condition.uniqueIdentifier);
}

- (void)testHighlightMatchConditionMutableCopyAndUniqueCopy
{
	IRCHighlightMatchCondition *original =
		[[IRCHighlightMatchCondition alloc] initWithDictionary:@{@"matchKeyword" : @"ping"}];
	IRCHighlightMatchConditionMutable *mutableCopy = [original mutableCopy];
	mutableCopy.matchKeyword = @"pong";
	mutableCopy.matchIsExcluded = YES;

	XCTAssertEqualObjects(original.matchKeyword, @"ping");
	XCTAssertFalse(original.matchIsExcluded);
	XCTAssertEqualObjects(mutableCopy.matchKeyword, @"pong");
	XCTAssertTrue(mutableCopy.matchIsExcluded);

	IRCHighlightMatchCondition *unique = [original uniqueCopy];
	XCTAssertEqualObjects(unique.matchKeyword, @"ping");
	XCTAssertNotEqualObjects(unique.uniqueIdentifier, original.uniqueIdentifier);
}

#pragma mark - Server

- (void)testServerDefaultsAndDictionaryRoundTrip
{
	IRCServer *server = [[IRCServer alloc] initWithDictionary:@{
		@"serverAddress" : @"irc.example.test",
		@"serverPort" : @6697,
		@"prefersSecuredConnection" : @YES
	}];

	XCTAssertEqualObjects(server.serverAddress, @"irc.example.test");
	XCTAssertEqual(server.serverPort, 6697);
	XCTAssertTrue(server.prefersSecuredConnection);
	XCTAssertGreaterThan(server.uniqueIdentifier.length, 0);

	IRCServer *empty = [IRCServer new];
	XCTAssertEqual(empty.serverPort, 6667);
	XCTAssertEqualObjects(empty.serverAddress, @"");
}

- (void)testServerMutablePasswordAndUniqueCopy
{
	IRCServerMutable *mutable = [IRCServerMutable new];
	mutable.serverAddress = @"chat.example.test";
	mutable.serverPort = 6667;
	mutable.serverPassword = @"s3cret";
	mutable.prefersSecuredConnection = NO;

	XCTAssertEqualObjects(mutable.serverPassword, @"s3cret");

	IRCServer *unique = [mutable uniqueCopy];
	XCTAssertEqualObjects(unique.serverAddress, @"chat.example.test");
	XCTAssertNotEqualObjects(unique.uniqueIdentifier, mutable.uniqueIdentifier);
	XCTAssertEqualObjects(unique.serverPassword, @"s3cret");
}

#pragma mark - Highlight log entry

- (void)testHighlightLogEntryMutableStoresLineClientAndChannel
{
	TVCLogLineMutable *line = [TVCLogLineMutable new];
	line.messageBody = @"hello world";
	line.nickname = @"alice";
	line.lineType = TVCLogLineTypePrivateMessage;
	line.receivedAt = [NSDate dateWithTimeIntervalSince1970:1700000000];

	IRCHighlightLogEntryMutable *entry = [IRCHighlightLogEntryMutable new];
	entry.lineLogged = line;
	entry.clientId = @"client-a";
	entry.channelId = @"channel-b";

	XCTAssertEqualObjects(entry.clientId, @"client-a");
	XCTAssertEqualObjects(entry.channelId, @"channel-b");
	XCTAssertEqualObjects(entry.lineNumber, line.uniqueIdentifier);
	XCTAssertEqualObjects(entry.timeLogged, line.receivedAt);
	XCTAssertGreaterThan(entry.renderedMessage.length, 0);
	XCTAssertGreaterThan(entry.timeLoggedFormatted.length, 0);
}

#pragma mark - Persistent store

- (void)testUserPersistentStoreHoldsRelationsAndTimerSlot
{
	IRCUserPersistentStore *store = [IRCUserPersistentStore new];
	IRCUserRelations *relations = [IRCUserRelations new];

	store.relations = relations;
	store.presentAwayMessageFor301LastEvent = 12.5;

	XCTAssertEqual(store.relations, relations);
	XCTAssertEqual(store.presentAwayMessageFor301LastEvent, 12.5);
	XCTAssertNil(store.removeUserTimer);
}

@end

NS_ASSUME_NONNULL_END

File: Tests/GlasstualTests/IRCModelMigrationTests.swift
/* *********************************************************************
 * Copyright (c) 2026 Codeux Software, LLC & respective contributors.
 * Please see Acknowledgements.pdf for additional information.
 *********************************************************************** */

import AppKit
import CocoaExtensions
@testable import Glasstual
import GlasstualPluginKit
import Testing

@MainActor
@Suite("IRC model migration")
struct IRCModelMigrationTests {
	@Test("A connection copies the config it was made with and starts idle")
	func connectionInitialStateAndConfigIsolation() {
		let client = GLTTestClient()
		var sourceConfig = Glasstual.IRCConnectionConfig()
		sourceConfig.serverAddress = "irc.example.test"

		let connection = Connection(config: sourceConfig, onClient: client)
		sourceConfig.serverAddress = "changed.example.test"

		#expect(connection.client === client)
		#expect(connection.config.serverAddress == "irc.example.test")
		#expect(connection.uniqueIdentifier.isEmpty == false)
		#expect(connection.isConnecting == false)
		#expect(connection.isConnected == false)
		#expect(connection.isDisconnecting == false)
		#expect(connection.isSecured == false)
		#expect(connection.isSending == false)
		#expect(connection.EOFReceived == false)
	}

	@Test("Resetting a connection clears every transient flag and the connected address")
	func connectionResetClearsTransientState() {
		let connection = Connection(config: Glasstual.IRCConnectionConfig(), onClient: GLTTestClient())

		connection.resetState()

		#expect(connection.isConnecting == false)
		#expect(connection.isConnected == false)
		#expect(connection.isConnectedWithClientSideCertificate == false)
		#expect(connection.isDisconnecting == false)
		#expect(connection.isSecured == false)
		#expect(connection.isSending == false)
		#expect(connection.EOFReceived == false)
		#expect(connection.connectedAddress == nil)
	}

	@Test("A seeded channel config takes the declared defaults, and the legacy keys still decode")
	func channelConfigDefaultsAndLegacyKeys() throws {
		let defaults = ChannelConfig.seed(withName: "#general")

		#expect(defaults.autoJoin)
		#expect(defaults.pushNotifications)
		#expect(defaults.showTreeBadgeCount)
		#expect(defaults.type == .channel)
		#expect(defaults.channelName == "#general")
		#expect(defaults.uniqueIdentifier.isEmpty == false)

		let legacy = try #require(PropertyListModel.decode(ChannelConfig.self, from: [
			"channelName": "#legacy",
			"joinOnConnect": false,
			"ignoreJPQActivity": true,
			"enableNotifications": false,
			"enableTreeBadgeCountDrawing": false,
		]))

		#expect(legacy.autoJoin == false)
		#expect(legacy.ignoreGeneralEventMessages)
		#expect(legacy.pushNotifications == false)
		#expect(legacy.showTreeBadgeCount == false)
	}

	@Test("A plain copy keeps the identifier while a unique copy takes a new one")
	func channelConfigCopiesAndUniqueCopiesPreserveValues() {
		var config = ChannelConfig(channelName: "#swift")
		config.label = "Swift migration"
		config.defaultModes = "+nt"
		config.secretKey = "join-key"

		let copy = config
		let unique = config.uniqueCopy()

		#expect(copy.channelName == "#swift")
		#expect(copy.label == "Swift migration")
		#expect(copy.defaultModes == "+nt")
		#expect(copy.secretKey == "join-key")
		#expect(copy.uniqueIdentifier == config.uniqueIdentifier)
		#expect(unique.channelName == "#swift")
		#expect(unique.secretKey == "join-key")
		#expect(unique.uniqueIdentifier != config.uniqueIdentifier)
	}

	@Test("A per-channel notification override is three-state, not a boolean")
	func channelConfigNotificationOverridesUseThreeStateSemantics() {
		var config = ChannelConfig()
		let event = TXNotificationType.highlight

		#expect(config.notificationEnabled(forEvent: event) == .mixed)

		config.setNotificationEnabled(.on, forEvent: event)
		#expect(config.notificationEnabled(forEvent: event) == .on)

		config.setNotificationEnabled(.off, forEvent: event)
		#expect(config.notificationEnabled(forEvent: event) == .off)

		config.setNotificationEnabled(.mixed, forEvent: event)
		#expect(config.notificationEnabled(forEvent: event) == .mixed)
	}

	@Test("A highlight condition decodes the legacy channel key and encodes it back")
	func highlightMatchConditionRoundTripsDictionaryAndDefaults() throws {
		let condition = try #require(PropertyListModel.decode(HighlightMatchCondition.self, from: [
			"matchKeyword": "alert",
			"matchChannelID": "chan-1",
			"matchIsExcluded": true,
		]))

		#expect(condition.matchKeyword == "alert")
		#expect(condition.matchChannelId == "chan-1")
		#expect(condition.matchIsExcluded)
		#expect(condition.uniqueIdentifier.isEmpty == false)

		let dictionary = PropertyListModel.encode(condition)

		#expect(dictionary["matchKeyword"] as? String == "alert")
		#expect(dictionary["matchChannelID"] as? String == "chan-1")
		#expect(dictionary["matchIsExcluded"] as? Bool == true)
		#expect(dictionary["uniqueIdentifier"] as? String == condition.uniqueIdentifier)
	}

	@Test("Editing a highlight condition leaves the value it was copied from alone")
	func highlightMatchConditionEditsAndUniqueCopyAreIndependent() {
		let original = HighlightMatchCondition(matchKeyword: "ping")
		var edited = original
		edited.matchKeyword = "pong"
		edited.matchIsExcluded = true

		#expect(original.matchKeyword == "ping")
		#expect(original.matchIsExcluded == false)
		#expect(edited.matchKeyword == "pong")
		#expect(edited.matchIsExcluded)

		let unique = original.uniqueCopy()

		#expect(unique.matchKeyword == "ping")
		#expect(unique.uniqueIdentifier != original.uniqueIdentifier)
	}

	@Test("A server decodes the keys it was given and defaults the rest")
	func serverDefaultsAndDictionaryRoundTrip() throws {
		let server = try #require(PropertyListModel.decode(Server.self, from: [
			"serverAddress": "irc.example.test",
			"serverPort": 6697,
			"prefersSecuredConnection": true,
		]))

		#expect(server.serverAddress == "irc.example.test")
		#expect(server.serverPort == 6697)
		#expect(server.prefersSecuredConnection)
		#expect(server.uniqueIdentifier.isEmpty == false)

		let empty = Server()

		#expect(empty.serverPort == 6667)
		#expect(empty.serverAddress == "")
	}

	@Test("A unique copy of a server carries the password over under a new identifier")
	func serverPasswordAndUniqueCopy() {
		var server = Server(serverAddress: "chat.example.test", serverPort: 6667)
		server.serverPassword = "s3cret"

		#expect(server.serverPassword == "s3cret")

		let unique = server.uniqueCopy()

		#expect(unique.serverAddress == "chat.example.test")
		#expect(unique.uniqueIdentifier != server.uniqueIdentifier)
		#expect(unique.serverPassword == "s3cret")
	}

	@Test("A highlight log entry records the line it was made from and where it came from")
	func highlightLogEntryStoresLineClientAndChannel() {
		let line = LogLine()
		line.messageBody = "hello world"
		line.nickname = "alice"
		line.lineType = .privateMessage
		line.receivedAt = Date(timeIntervalSince1970: 1_700_000_000)

		let entry = HighlightLogEntry(lineLogged: line, clientId: "client-a", channelId: "channel-b")

		#expect(entry.clientId == "client-a")
		#expect(entry.channelId == "channel-b")
		#expect(entry.lineNumber == line.uniqueIdentifier)
		#expect(entry.timeLogged == line.receivedAt)
		#expect(entry.renderedMessage.length > 0)
		#expect(entry.timeLoggedFormatted.isEmpty == false)
	}

	@Test("A user's persistent store keeps its relations and leaves the timer slot empty")
	func userPersistentStoreHoldsRelationsAndTimerSlot() {
		let store = UserPersistentStore()
		let relations = UserRelations()
		store.relations = relations
		store.presentAwayMessageFor301LastEvent = 12.5

		#expect(store.relations === relations)
		#expect(store.presentAwayMessageFor301LastEvent == 12.5)
		#expect(store.removeUserTimer == nil)
	}
}
